Pause / Resume Crawls Initial Implmentation.

ikreymer opened the pull request:
Work on #2566 
Fixes #2576 

- add 'pause' crawl state (fixes #2567)
- turns off crawler pods, and then redis pod when paused
- crawler uploads WACZ before shutting down (dependent on webrecorder/browsertrix-crawler#824)
- add 'paused_at' on crawl spec to indicate when crawl is paused
- support max pause time limit, after which crawl becomes automatically stopped.
- add 'stopped_pause_expired' when pause automatically expires and crawl is stopped
- /crawl/<id>/{un}pause apis to toggle 'paused' on crawl spec
- ui: add pause/resume button, paused state (partially addresses #2568)
- ui: add pausing/unpausing derivative states when crawl is running and pausing, or paused and not pausing (partially addresses #2569)

Todo:
- How to handle crawler versions that don't auto-upload WACZ / don't respond to pause message

ikreymer committed:
pause/resume initial impl:
- add 'pause' crawl state
- turns off crawler pods, and then redis pod when paused
- add 'paused' on crawl spec to indicate when crawl is paused
- /crawl/<id>/{un}pause apis to toggle 'paused' on crawl spec
- ui: add pause/resume button, paused state
- ui: add pausing/unpausing derivative states when crawl is running and pausing, or paused and not pausing

tw4l committed:
add auto pause expiry support:
- stop crawls that have been paused for too long
- add 'paused_crawl_limit_minutes' to Helm chart
- add paused time and expiry to crawlconfig API response
- set to 'stopped_pause_expired' state
- ui: add support for 'Stopped: Paused Too Long' for stopped_pause_expired
- use 'paused_at' in CrawlJob to indicate crawl is paused and when

SuaYoo committed:
Pause / resume frontend tweaks (#2575)

- Hide "Pause" button if it's not relevant, instead of disabling without
a displayed reason
- Make "Resume" button primary
- Use circular icons to match other status icons
- Show toast message on successful pause/unpause

ikreymer committed:
paused redis flag:
- set <crawlid>:paused key when a crawl is paused and at least one crawler pod exists
- clear <crawlid>:paused when crawl is paused and more pods running
ensure flag is cleared before redis is shutdown, already cleared when a crawl is unpaused

ikreymer committed:
support stopping paused crawls instantly (if WACZ files already uploaded)

ikreymer committed:
refactor to use existing :stopping approach to allow working with previous versions of crawler:
- set :stopping key to true
- when crawler pod exits, immediately reset done -> interrupted
- clear :stopping key when all crawler pods have exited OR crawl not longer paused (to allow resume)

ikreymer committed:
pass 'pausing' to btrix-crawl-status (and also a missing 'stopping')

SuaYoo committed:
darker gray for paused state

SuaYoo commented on frontend/src/pages/org/workflow-detail.ts:
```diff
@@ -616,12 +616,37 @@ export class WorkflowDetail extends BtrixElement {
+      this.workflow.lastCrawlStopping;
+    const disablePause =
+      this.workflow.lastCrawlPausing ===
+      (this.workflow.lastCrawlState === "running");
```
Should this just be checking `lastCrawlPausing`?
```suggestion
    const disablePause = this.workflow.lastCrawlPausing;
```

ikreymer replied:
I think the idea was that the button is both Pause / Resume, and possible states are:
- pausing and running (disable, already in process of Pausing)
- pausing and not running (don't disable, it should show Resume)
- not pausing and not running (disable, already in process of Resuming)
- not pausing and running (don't disable, it should show Pause)

Had to double check that, perhaps a bit more confusing then it needs to be.. maybe should rename it to disablePauseResume to be more clear

ikreymer replied:
The `lastCrawlPausing` / `pausing` reflects if user requested crawl to be paused/not paused, while crawl state is the actual state the crawl is in, so we enable the button when they are different.

SuaYoo replied:
Yeah a variable rename would work, `hidePauseResume` and `disablePauseResume`.

Is there any need to add and check for `lastCrawlResuming`?

ikreymer replied:
> Yeah a variable rename would work, `hidePauseResume` and `disablePauseResume`.
> 
> Is there any need to add and check for `lastCrawlResuming`?

Resuming is the when lastCrawlPausing is set to false. I suppose it could be renamed too - if true, it means the user wants the crawl paused, if false, it means the user wants the crawl running/resumed (the default).

ikreymer replied:
Updated this to hidePauseResume and disablePauseResume. Also did a pass to replace instance of 'unpause' with 'resume', including in the API, to be consistent.

ikreymer committed:
rename for clarity, use resume instead of unpause consistently

tw4l committed:
Add pausedExpiryMinutes to /settings API endpoint

ikreymer committed:
rename pausing -> shouldPause, lastCrawlPausing -> lastCrawlShouldPause, except actual derived 'pausing' state in frontend
ensure resources are available for paused state

ikreymer committed:
undo previous changes, just need to check for 'state' being paused

ikreymer committed:
operator: refactor pausing under assumption crawl pod stays in 'completed' after being
paused, goes to running state update
- set state to 'pause' only when all crawler pods are paused
- log 'pausing' on first change, don't recheck stopping if already set

ikreymer committed:
use :paused key again, support crawler 1.6.1+, otherwise duplicate data is uploaded

ikreymer commented:
Had to do some additional refactoring, switching back to using :paused key, supporting only newer crawler versions, otherwise duplicate data is uploaded. Also cleans up the implementation a bit.

ikreymer committed:
ui: add 'shouldPause' to admin running crawls display
